Priority Scheduler Feature Example

sreid8 opened the pull request:
Per conversation with jentfoo, examples needed to be created for each of the [features on the threadly wiki.](https://github.com/threadly/threadly/wiki/Threadly-Features) The first of these to be attempted is the [PriorityScheduler](https://github.com/threadly/threadly/blob/master/src/main/java/org/threadly/concurrent/PriorityScheduler.java).

The example takes a number of threads as an argument. It creates that number of threads and tasks (simple counting task) and randomly assigns priorities to the tasks. All threads are then started at the same time and time from the start of the thread to the end of thread execution is reported when the thread terminates. "-v" can be passed to get additional detail about the status of execution of each thread.

Comments?

jentfoo commented on src/main/java/org/threadly/examples/features/PrioritySchedulerExample.java:
```diff
@@ -0,0 +1,164 @@
+   * Main entry point of the example.
+   * @param args -> program args. run w/o args for help
+   */
+  public static void main(String[] args) {
```
Does it make sense to have these feature examples able to be ran?  I had not considered it, but thinking about it now, I am thinking maybe not?  These don't seem interesting to run, and the code around execution just distracts from what we are really trying to demonstrate.

Let me know your thoughts, but my vote is to axe the main function and such.  For actual development, feel free to add /src/test/java/... and put unit tests in which can start the execution.  Don't worry about coverage, I am just mentioning that unit test could be a way to test execution of the feature example without a main function.  If you don't feel the unit tests are necessary, feel free to not include any (for this example work).

jentfoo commented on src/main/java/org/threadly/examples/features/PrioritySchedulerExample.java:
```diff
@@ -0,0 +1,164 @@
+    //add numThreads - 1 tasks to the PriorityScheduler.
+   for (int i = 0; i <= numThreads; i++) {
+     //random between 0-2 to determine priority.
+     TaskPriority priority = getRandomPriority();
```
I was thinking in these examples we would do things which are more likely to be real use cases.  I can't think of a use case where a random priority would be helpful.  So for example I would recommend this as an alternative:

Submit some tasks with priority, but have no implementation, instead just a comment describing what they would do.  Either as inner classes, or anonymous inner classes, some examples we may want to show:
* A recurring low priority task that checks a source for configuration...we could describe how this task could be checking for a configured pool size, and updating the pool size at runtime if desired
* A recurring low priority task that does some cleanup work...possibly a cache cleaner, or something similar (again no implementation, just a comment describing a use case)...it might also be worth constructing the priority scheduler with a longer non-default low priority wait time, and describe that (in our silly example) low priority tasks can wait for say 5 or 10 seconds beyond scheduled run time without any impact in the system
* A recurring high priority task that reports statistics.  One possible stat it could be reporting is how many tasks are queued in the pool, and how many tasks are running.
* Maybe have some high priority executions, which could either come in from a public function (that is never invoked), or something like this loop

jentfoo commented on src/main/java/org/threadly/examples/features/PrioritySchedulerExample.java:
```diff
@@ -0,0 +1,164 @@
+          System.out.println("Thread " + runNumber + " is on iteration " + (i) + ". ");
+        }
+      }
+      System.out.println("Thread " + runNumber + " took " + (System.nanoTime() - startTime) + "  nanoseconds to execute." );
```
Instead of System.nanoTime, check out the Clock class: http://threadly.github.io/threadly/javadocs/4.3.0/org/threadly/util/Clock.html

We should have another example class for that, so no need to go deeply into it, but we might as well eat our own dog food in our examples.

jentfoo commented on src/main/java/org/threadly/examples/features/package-info.java:
```diff
@@ -0,0 +1,8 @@
+ * @author sreid8
+ *
+ */
+package org.threadly.examples.features;
```
I don't like to have the author tags in the package javadocs.  Could you add a small one or two sentence description here, for example:

A collection of small examples using specific threadly features.

jentfoo commented:
In general things look really good, thanks for your help in this.  Take a look at my comments and let me know if you disagree with any.  Otherwise once those are addressed I think this can be merged.

sreid8 commented:
I've modified the example to work more with how I understand the comments above. Does it make sense to add a class to run this example or should use be able to be gleaned from the methods in the example? Any further suggestions?

jentfoo commented on src/main/java/org/threadly/examples/features/PrioritySchedulerExample.java:
```diff
@@ -0,0 +1,146 @@
+         * This is the Starvable priority task that will be run.
+         * Starvable tasks run only when no other task needs to run
+         * It can be used to keep track of configuration:
+         * Example use case: closing sockets that no longer have hosts
```
This example use case does not make sense for starvable, because if it was actually starved the sockets would end up never closing and you would run out of file handles.

Right now I don't have a an amazing use case for starvable.  The best I have considered is a starvable task could be used as a indication to bring in more work.  Not like pulling from a queue (since usually the task would do that), but maybe it calls out to a micro service to indicate that it is ready for an upstream service to produce more work for it.

I would be okay with ignoring the starvable example if we want for now.

jentfoo commented on src/main/java/org/threadly/examples/features/PrioritySchedulerExample.java:
```diff
@@ -0,0 +1,146 @@
+        /*
+         * This is the High priority task that will be run.
+         * High priority tasks are the highest priority in the thread pool.
+         * Example use case: Stats reporting about number of threads active in the pool
```
I think the best example of high priority tasks is servicing the actual requests of an application, or doing processing work where the result needs to be produced as soon as possible.

jentfoo commented on src/main/java/org/threadly/examples/features/PrioritySchedulerExample.java:
```diff
@@ -0,0 +1,146 @@
+         * Low priority tasks are run before Starvable tasks,
+         * but exactly when it will be run is enforced by the 
+         * PrioritySchedulerService.
+         * Example use case: cache cleaning or other cleanup work
```
I would make the low priority example a recurring task (scheduleAtFixedDelay), you could save a reference to the task and add a "shutdown" function which removes the recurring task from the pool (thus showing our guaranteed task removal, which on java.util does NOT exist)

sreid8 replied:
The only thing I'm confused about is the shutdown piece. Where would the shutdown method be implemented and how is it registers such that the task will be interrupted when the PriorityScheduler is stopped? It seems that the shutdown and shutdownNow methods of the PriorityScheduler (according to javadoc, at least) specifically does not interrupt any running tasks... so how would the shutdown method be executed in such a way as to interrupt the recurring task?

jentfoo replied:
Ah, sorry, I was a bit confusing there.

I was thinking you would implement a "dispose" or "shutdown" that would just remove the recurring runnable...but I can see how that would be a little awkward since you are constructing the priority scheduler (it would make more sense if it was passed in).  The idea being "I constructed this, it ran for a while, and now I am no longer interested in it, so stop it, and let me use my resources for something else".

No matter the case it would not send an interrupt to the running task.  It would just prevent future executions.

Needless to say this is example code, so I don't know how much sense it needs to make or not.  I am fine if you want to omit it.

jentfoo commented:
I added some small details as to what I think would be more likely and identifiable use case examples.  That said I think this looks a ton closer to what I wanted!  I don't think it makes sense to make it actually run, I don't see any value in it (but if you want to add another class to run it, it wont hurt, just not important to me).

I still want to link these examples from the wiki once we merge them in, so they should be easy to find and see.

If you want to take care of those last few things I mentioned I am happy to merge this.

sreid8 commented:
I added a dispose method that I believe works correctly based on how I understand the javadocs on shutdown and shutdownNow for the PriorirtyScheduler. I've also added a truly recurring task to the task generation method and parameterized the task generation method.

jentfoo commented on src/main/java/org/threadly/examples/features/PrioritySchedulerExample.java:
```diff
@@ -0,0 +1,153 @@
+      public void run() {
+        /*
+         * This is the task that will be run on a schedule.
+         * Example use case: A recurring task
```
Sorry for nit picking, I can fix this if you want.  Example use case "a recurring task" is not very informative.  I think "Example use case: regular cleanup/system maintenance task"

jentfoo commented on src/main/java/org/threadly/examples/features/PrioritySchedulerExample.java:
```diff
@@ -0,0 +1,153 @@
+   */
+  public static void disposePrioritySchedulerTasks(final PriorityScheduler executor, boolean recurringStop) {
+    if (recurringStop) {
+      executor.shutdownNow();
```
I think what you meant to do here here is remove the recurring task.  In order to do that you should make this non-static, and save a reference to the runnable your constructing above (and submitting at a fixed rate).  You should then call .remove(Runnable) to prevent future executions.

sreid8 commented:
No problem with the nitpicking. I'd rather contribute usefully anyway

I also attempted to implement what I think you were hoping for with the dispostPrioritySchedulerTasks() method suggestions. If I'm off the mark, I'll definitely take another pass at it.

jentfoo commented:
LGTM, thanks for all your work on this!  I will go ahead and merge this.  Would you like to update the wiki to link to this code example, or would you prefer I do that?

sreid8 commented:
I'll update it this weekend.
